アクセストークン発行用のアクションを追加

a5ob7r opened the pull request:
# 概要

OAuth 2.0のResource Owner Password Credentials Grantを利用してアクセストークンを発行するアクションを追加する。

- usernameパラメータの値のメールアドレスが送信されることを想定している。
    - このパラメータはOAuth 2.0のRFCにおいて、アクセストークンを発行する際に必須となるパラメータであり、その内容については特に規定されていない。IDとして機能するのであれば仕様上は問題ないと思われる。
- エラーレスポンスについては、返すべき状況とそのパラメータがRFCに記載されている。
    - https://datatracker.ietf.org/doc/html/rfc6749#section-5.2

a5ob7r committed:
Add a POST action to issue an access token

a5ob7r committed:
Add 500 as a error response to the API document

a5ob7r committed:
Fix a description for the username parameter

This parameter must be a user's email address to identify the user.

a5ob7r committed:
Add two headers to disable caching for access token issuing

a5ob7r committed:
Remove some duplicate tests

We make emphasis on the difference between whether the scope is included
or not by this change.

tim-toga commented on openapi.yaml:
```diff
@@ -347,7 +364,7 @@ paths:
                 username:
                   type: string
-                  description: アクセストークンの要求に利用するリソースオーナーのユーザ名である。
+                  description: アクセストークンの要求に利用するリソースオーナーのメールアドレスである。ユーザの表示名ではないことに注意する。
```
> ユーザの表示名ではないことに注意する。

と書いてくれてる通り、usernameパラメータにメールアドレスを渡すのは分かりにくいです。emailパラメータとかにリネームするとダメな理由があるんですか？

a5ob7r replied:
OAuth 2.0のRFCにてパラメータ名が指定されているためです。

https://datatracker.ietf.org/doc/html/rfc6749#section-4.3.2

tim-toga replied:
それはusernameパラメータが必要な理由じゃないです？
usernameパラメータにメールアドレスを渡すのは、クライアントのミスを誘発させると思うので避けたいですね

a5ob7r replied:
あとは、APIの仕様書でユーザのname属性を一意にしていないということも理由の一つです。他にこのusernameに渡す適切な、つまりは一意な値がないんですよね。一意というのなら、IDも当てはまりはするのですが...。

それ以外だと、usernameの値としてメールアドレスを渡している例をいくつか見かけたのも、理由の一つではあります。

> ...
username=testuser@salesforce.com&
...

https://help.salesforce.com/s/articleView?id=sf.remoteaccess_oauth_username_password_flow.htm&type=5
https://learn.microsoft.com/ja-jp/azure/active-directory/develop/v2-oauth-ropc

a5ob7r replied:
無論、直感的ではないことは承知しています。API設計書を書いたのに、このことを失念していたので、反省するところではあります。

a5ob7r replied:
まあ、現状ではメールアドレスを登録させてはいますが、現状ではユーザが自由に変更可能なIDというくらいの利用法しかないです。別にユーザにメールアドレスを送るわけではないので、メールアドレスを除去してIDとして機能する一意なusername属性を追加するというのもありかもしれないですね。メールアドレスを残してもいいですけど、全く利用する予定がないので、IDとして利用しないのであれば、残しておく理由が特にないです。

あとは、本API自体はトレーニング目的だし、今回のことは教訓として胸にとどめておけばいいか、と楽観的に考えていました。

tim-toga replied:
なるほどー。usernameにメールアドレスを渡すのがベターなのかー

tk0miya replied:
OAuth との整合させるためには仕方ない部分じゃないかな

tk0miya commented on .rubocop.yml:
```diff
@@ -20,6 +20,30 @@ AllCops:
+    - app/controllers/**/*
+
+Metrics/MethodLength:
+  Max: 12
```
雑談: ここはデフォルト値が結構厳しいんですよね。多くのケースではもっと広げてます。

tk0miya commented on app/controllers/access_tokens_controller.rb:
```diff
@@ -0,0 +1,45 @@
+
+class AccessTokensController < ApplicationController
+  def post
+    render problem: { error: 'invalid_request' }, status: :bad_request and return unless contains_required_params?
+    render problem: { error: 'unsupported_grant_type' }, status: :bad_request and return unless supported_grant_type?
+    render problem: { error: 'invalid_grant' }, status: :bad_request and return if current_user.nil?
```
この手の事前条件は before_action で書くことがおおいですね

* そのアクションを実行するための事前条件 (ログインしていること、トークンが正しいこと、など) は before_action
* そのアクションの実装、本体はアクションそのもの

tk0miya replied:
レビューをすすめて気づいたけど、最後の行(invalid_grant)は ID, passwd で認証するというこのアクションの重要な処理なので、事前条件にしてはまずいやつですね。上2行は事前条件っぽい。

a5ob7r replied:
> この手の事前条件は before_action で書くことがおおいですね

あー、そういえばそういうものがありましたね。失念していました。

tk0miya commented on app/controllers/access_tokens_controller.rb:
```diff
@@ -0,0 +1,45 @@
+  private
+
+  def current_user
+    @current_user ||= User.find_by(email: params[:username])&.authenticate params[:password]
```
メモ: `#authenticate` は self を返す
https://github.com/rails/rails/blob/cdd14ce1f5196e4bd98df42f89a1cd36ba9d4bee/activemodel/lib/active_model/secure_password.rb#L143-L148

※ 直感的ではないので `#current_user` の中で `#authenticate` を呼ぶのはちょっと悩ましい。メソッド名から認証している感が湧かないのですよね

tk0miya replied:
`#current_user` ではなくて、 `#authenticate` とか `#login` とか、そういう名前なら問題ないのかもなあ。

a5ob7r replied:
#current_userを削除し、#post内で直接ユーザの探索と認証を行うようにしました。こちらで指摘されたアーリーリターンを3つ並べたかったので、メソッドとして切り出していましたが、そのうち2つを.before_actionで実行するようにしたので、アーリーリターンを並べることを考える必要がなくなりました。なので、こちらもメソッドとして切り出す必要性がなくなりました。

https://github.com/sls-training/2023-API-training-server/pull/38#discussion_r1243131470

こういう重要な処理はアクション内で直接実行させたほうがいいですかね。複雑というわけでもないですし。

tk0miya replied:
はい、同意です。"before" ではなく "main" になるべき処理ですよね。
複雑であったとしてもアクションで実行すべきだと思います。
呼び出しが複雑なので、メソッドに切り出すことはあっても、before_action に移動すべきではないです。
その上で、このくらいの記述であればメソッドに切り出すほどではないというのはどういうです。

tk0miya commented on app/controllers/access_tokens_controller.rb:
```diff
@@ -0,0 +1,45 @@
+  private
+
+  def current_user
+    @current_user ||= User.find_by(email: params[:username])&.authenticate params[:password]
+  end
+
+  def access_token_params
+    params.permit :scope
+  end
+
+  def contains_required_params?
+    params.keys.to_set.superset? %w[grant_type username password].to_set
+  end
```
提案: 僕ならどう書くだろう? と考えてみた。文字数はちょっと多いけど、やりたいことをストレートに表現できている気がするがどうだろうか。

```suggestion
  def contains_required_params?
    params.require(%w[grant_type username password])
  rescue ActionController::ParameterMissing
    false
  end
```

a5ob7r replied:
なるほど、ありですね。

a5ob7r replied:
集合を利用した方法は確かにわかりづらかったので、#all?を利用する方法に変更しました。今回はこれで行こうと思います。

#requireを利用する場合、このメソッドはキーの存在のみだけでなく、値が#present?かどうかも見てくれるようなので、それが必要であれば、そちらを利用するかもしれませんね。でも、他に方法がないならともかく、ここで例外を使うのはなんだかなー、と少しだけモヤモヤを感じています。
https://api.rubyonrails.org/classes/ActionController/Parameters.html#method-i-require

tk0miya replied:
`#all?` を使った今のコードも十分にシンプルでよさそうです。

tk0miya commented on config/routes.rb:
```diff
@@ -1,3 +1,4 @@
 Rails.application.routes.draw do
   post 'user', to: 'users#post'
+  post 'signin', to: 'access_tokens#post'
```
雑談より: Rails は RESTful な URL 設計と強く結びついているので、RESTful ではない構造にすると get, post などの低レイヤーのルーティングが必要になりますね。
悪手であるとまでは言わないものの、resource で表現できないものが出てきた場合は嫌な匂いではないか振り返ってみると良さそう。

※ 今回は API デザインの時点で `/signin` を選んでいるので、意図的ですよね。

tk0miya commented on app/controllers/access_tokens_controller.rb:
```diff
@@ -0,0 +1,45 @@
+    params[:grant_type] == 'password'
+  end
+
+  def set_disble_caching_headers
```
雑談: ここでやっていることって、

* コード上では「キャッシュを無効化するヘッダを設定する」こと
* 全体的には「キャッシュを無効化する」こと

ですよね。
手段じゃなくて目的に着目して `#disable_caches` という名前にしてもよいのでは…とちょっと頭をよぎったのだけど、それだと「なんのキャッシュを無効化するの?」とか別の雑念が生まれそうなので、まあこのままでいいかな、と考えました。

tk0miya commented on openapi.yaml:
```diff
@@ -311,6 +311,23 @@ components:
                   "error": "insufficient_scope",
                   "scope": "write"
                 }
+    internalServerError:
+      description: リクエスト自体は正しいが、サーバ上で何らかのエラーが発生し処理ができなかったことを意味する。
```
500 エラーを定義するかどうかは悩ましいなあ。

500 はその他何かしらのサーバエラーなので、これが返るときはレスポンスは約束できないしね。
たとえば DB にアクセスできない場合、メモリ不足の場合なども 500 エラーになりますね。
そのときに、レスポンスはここに記載したものは返せないですし

* 意味のあるエラーを返すのであれば、500 以外を使うのがいいと思う
* なにか上手く扱えなかった、という意味では 500 は周知のエラーなので定義しなくて良い

という風に考えます。

a5ob7r replied:
なるほど、そういうものなんですね。

a5ob7r replied:
確かにそのような共通認識がありそうですね。今回は該当する記述を削除することにします。

https://stackoverflow.com/questions/54989351/should-internal-server-error-be-documented-in-swagger

tk0miya commented on openapi.yaml:
```diff
@@ -365,6 +382,17 @@ paths:
             ログインに成功してアクセストークンが発行されたことを意味する。
 
             https://datatracker.ietf.org/doc/html/rfc6749#section-5.1
+          headers:
+            'Cache-Control':
+              schema:
+                type: string
+                enum:
+                  - no-store
+            'Pragma':
+              schema:
+                type: string
+                enum:
+                  - no-cache
```
ここって OAuth とか、何かの規格で使うべきと指定されてるやつでしたっけ。
あまり API では使わないので、ちょっと気になりました。

* このあたりの仕組みはブラウザ向け
* API クライアントを作るときはわざわざ実装が必要なもの
* そもそも POST リクエストをキャッシュすることはほぼない

参考: https://developer.mozilla.org/ja/docs/Web/HTTP/Caching

a5ob7r replied:
> ここって OAuth とか、何かの規格で使うべきと指定されてるやつでしたっけ。

はい、そうです。

> The authorization server MUST include the HTTP "Cache-Control"
   response header field [[RFC2616](https://datatracker.ietf.org/doc/html/rfc2616)] with a value of "no-store" in any
   response containing tokens, credentials, or other sensitive
   information, as well as the "Pragma" response header field [[RFC2616](https://datatracker.ietf.org/doc/html/rfc2616)]
   with a value of "no-cache".

https://datatracker.ietf.org/doc/html/rfc6749#section-5.1

a5ob7r replied:
APIの利用において役立つ情報ではないので、該当する記述を削除しました。

tk0miya commented on spec/requests/access_tokens_spec.rb:
```diff
@@ -0,0 +1,108 @@
+        expect(subject.get_header('Pragma')).to eq 'no-cache'
+      end
+
+      it 'レスポンスボディに発行したアクセストークンとそのメタデータを含む' do
```
これはパラメータに scope が含まれていないときの example では? context 持たせなくていいんでしたっけ

a5ob7r replied:
あー、それもそうですね。修正します。

tk0miya commented on spec/requests/access_tokens_spec.rb:
```diff
@@ -0,0 +1,108 @@
+
+      context 'パラメータにscopeが含まれているとき' do
+        subject do
+          post signin_path, params: { grant_type:, username:, password:, scope: }
```
subject をいくつも書きたくないので、僕ならこう書きますね。
```suggestion
          post signin_path, params: { grant_type:, username:, password:, scope: }.compact
```

こう書いておくと `scope` 変数にどういう値をセットするかで、パラメータを送信するのかしないのかを制御できます (subject の定義は１つだけで済む)

a5ob7r replied:
へー、こんなメソッドがあるんですね。

https://docs.ruby-lang.org/ja/latest/method/Hash/i/compact.html

tk0miya replied:
Array や Enumerable にも存在するので、ちょこちょこ使いますね。
この手のメソッドは徐々に手持ちのカードを増やしておくと便利なので、都度覚えていくと良さそうです。

a5ob7r committed:
Add some before_actions instead of validation in the action

a5ob7r committed:
Remove #current_user

Finding and authenticating users are parts of the main logic for this
action, so we should run it in the action directly.

a5ob7r committed:
Add a missing context

a5ob7r committed:
Remove the response for 500

Users probably get on the same page that servers may response with 5XX
status codes every time. So we don't need to indicate explicitly in the
API documents.
https://stackoverflow.com/questions/54989351/should-internal-server-error-be-documented-in-swagger

a5ob7r committed:
Remove some unnecessary Cops

No code violates these rules.

a5ob7r committed:
Remove cache related headers from the API document

Probably these description doesn't provide any additional and useful
information about these APIs.

a5ob7r committed:
Change how to check whether required parameters exist or not

Using Set is not very understandable.

a5ob7r commented:
修正が必要そうな箇所については対応しました。

a5ob7r commented:
ありがとうございました。
